### ISSUE TITLE
Netkit does not work with endpoint routes in some cases

### ISSUE THREAD

**Issue by jrife** [cilium/cilium#35060]
This tracks a group of issues related to netkit that share a similar root cause and discussion around it.

- In #34042 a user reported that liveness/readiness probes stopped working after applying `NetworkPolicy` to a `Pod` when using `netkit` and endpoint routes. The same configuration using `veth` did not have this problem. Some [investigation](https://github.com/cilium/cilium/issues/34042#issuecomment-2350351085) turned up that packets coming from the host namespace get misidentified with the `world` identity instead of `host`. With `NetworkPolicies` applied to the endpoint this leads to packets from liveness/readiness probes getting dropped.
- In #33875 a user [reported](https://github.com/cilium/cilium/issues/33875#issuecomment-2365156011) that running `curl -I -s https://api.github.com/` from a `Pod` hangs when using `netkit` and endpoint routes after applying a `CiliumNetworkPolicy` which specifies `.spec.egress.toFQDNs`. The same configuration using `veth` or with endpoint routes disabled did not have this problem. Some [investigation](https://github.com/cilium/cilium/issues/33875#issuecomment-2375774287) turned up a similar symptoms, with reply DNS traffic returning from the proxy taking on the `world` identity in `cil_to_container`.

In both cases the root cause looks to be that netkit scrubs packets (clearing `skb->mark`) before executing attached BPF programs. This is evident looking at the source for [`netkit_xmit`](https://github.com/torvalds/linux/blob/075dbe9f6e3c21596c5245826a4ee1f1c1676eb8/drivers/net/netkit.c#L66) where `netkit_prep_forward` clears `skb->mark` when crossing network namespaces before `netkit_run()` runs any attached programs.

```c
	netkit_prep_forward(skb, !net_eq(dev_net(dev), dev_net(peer)));
	eth_skb_pkt_type(skb, peer);
	skb->dev = peer;
	entry = rcu_dereference(nk->active);
	if (entry)
		ret = netkit_run(entry, skb, ret);
```

In contrast, when using `veth` TC/TCX hooks are executed *before* the `veth` driver does any packet scrubbing.

#### `netkit` egress processing order
1. [`sch_handle_egress`](https://github.com/torvalds/linux/blob/075dbe9f6e3c21596c5245826a4ee1f1c1676eb8/net/core/dev.c#L4115) is called but does not execute any hooks, since BPF programs are attached directly to the device in `netkit` mode.
2. `netkit_xmit` begins
3. `netkit_xmit` clears `skb->mark`
4. `netkit_xmit` runs `cil_to_container`
5. `netkit_xmit` passes the packet to the peer device

#### `veth` egress processing order
1. [`sch_handle_egress`](https://github.com/torvalds/linux/blob/075dbe9f6e3c21596c5245826a4ee1f1c1676eb8/net/core/dev.c#L4115) is called executing `cil_to_container`.
2. `veth_xmit` begins
3. `veth_xmit` clears `skb->mark`
4. `veth_xmit` passes the packet to the peer device

Since `cil_to_container` uses `ctx->mark` for proxy redirection and policy enforcement this can lead to various issues. For now this only seems to be a problem when using endpoint routes, but if `cilium_host` were to be changed to `netkit` (I'm not sure if this is the case already) it may interfere there as well.

Just to confirm that this is indeed the root cause, patching the netkit driver with this hack locally resolves the reported issues in both cases making behavior consistent between `veth` and `netkit` modes (note: this is not a real fix, just something I used to test my observations).

```
 static void netkit_prep_forward(struct sk_buff *skb, bool xnet)
 {
+       u32 save_mark = skb->mark;
        skb_scrub_packet(skb, xnet);
+       skb->mark = save_mark;
        skb->priority = 0;
        nf_skip_egress(skb, true);
        skb_reset_mac_header(skb);
```

## Tasks
- [ ] Short term, decide whether or not we should allow netkit to be used with endpoint routes enabled. We could add some checks on startup to block this combination. Cons: this might make things stop working for some users who are already using this combination without issue.
- [ ] Document this as a known issue (#35126)
- [ ] Patch netkit. Talking to @borkmann offline, he mentioned maybe adding a new mode to determine if the scrub happens before or after running BPF. If implemented, this needs some follow up work to make sure this mode is configured.
- [ ] Expand test coverage to include some of these scenarios with netkit+endpoint routes.

**Comment by jrife** [comment]
CC @julianwiedmann @joestringer who are also involved in the issues mentioned.

**Comment by jrife** [comment]
Just wanted to add, I'm happy to take a stab at adding a new mode to netkit if it would help.

**Comment by borkmann** [comment]
Hi @jrife I've pushed a patch here which I think could be acceptable. Do you have a chance to help testing whether that looks good on your side? https://github.com/cilium/linux/commits/nk-scrub/  I can run this by Martin and Nik in the meantime. Thanks for your help!

**Comment by borkmann** [comment]
> Short term, decide whether or not we should allow netkit to be used with endpoint routes enabled. We could add some checks on startup to block this combination. Cons: this might make things stop working for some users who are already using this combination without issue.

We could try asking for stable backport once this lands. I can add a Fixes tag just in case.

**Comment by jrife** [comment]
Cool, I'll give it a try locally under the two scenarios mentioned.

> We could try asking for stable backport once this lands. I can add a Fixes tag just in case.

That would be great. It looks like we'd need to backport a fix to all supported Cilium versions to use `NETKIT_SCRUB_AFTER` if supported by the current kernel. I wonder if there should be any logic in Cilium to block a user from using endpoint routes with netkit on an unpatched kernel? Then again, maybe a known issue is enough since affected users could just switch to `veth` until they update their kernel.

**Comment by jrife** [comment]
> I'll give it a try locally under the two scenarios mentioned.

I [changed Cilium locally](https://github.com/jrife/cilium/commit/0f8506c99310188292850918461e17048b0de8a8#diff-dc342c45d117c4ae7302bf54995f7a41ccfd33be4ff43286b89fa7b8295e13c5) to set `IFLA_NETKIT_SCRUB` to `NETKIT_SCRUB_AFTER`.

* I tested the scenario in #34042 (liveness probes + `NetworkPolicy`) with the kernel patch and it seems to fix the issue. Switching to `NETKIT_SCRUB_BEFORE` this breaks again as expected.
* The scenario in #33875 does not work still unless `NETKIT_SCRUB_NONE` is used. I believe this is related to the fact that `skb->mark` is immediately reset after running `cil_to_container()` which sets it in `__ctx_redirect_to_proxy`.
  ```c
  ctx->mark = MARK_MAGIC_TO_PROXY | proxy_port << 16;
  ```
  Packets disappear after being redirected to the proxy. I can take a bit of a closer look tomorrow.
  ```
  <- network flow 0x0 , identity unknown->unknown state unknown ifindex eth0 orig-ip 0.0.0.0: 10.244.0.156:53 -> 10.244.1.246:36813 udp
  -> proxy flow 0x0 , identity 10388->26256 state reply ifindex lxc7b7ce9de2500 orig-ip 10.244.0.156: 10.244.0.156:53 -> 10.244.1.246:36813 udp
  <- network flow 0x0 , identity unknown->unknown state unknown ifindex eth0 orig-ip 0.0.0.0: 10.244.0.156:53 -> 10.244.1.246:36813 udp
  -> proxy flow 0x0 , identity 10388->26256 state reply ifindex lxc7b7ce9de2500 orig-ip 10.244.0.156: 10.244.0.156:53 -> 10.244.1.246:36813 udp
  ```

**Comment by borkmann** [comment]
Hah, okay, thanks for testing and taking a look. I presume NETKIT_SCRUB_NONE generally could be enough as an option anyway and if needed BPF prog could zero the data. I'll finalize and add a BPF selftest (next week) so that this can be submitted & merged then.

**Comment by jrife** [comment]
To expand a bit on the behavior I saw in the second scenario for posterity,

1. Used `NETKIT_SCRUB_AFTER`
2. The coredns Pod sends a UDP packet back to the curl Pod.
3. `netkit_xmit()` starts and runs `netkit_run()` triggering `cil_from_container` for the coredns Pod.
4. `cil_from_container` calls `ipv4_local_delivery()` which tail calls to `handle_policy()` for the curl Pod.
5. Eventually this hits `ctx_redirect_to_proxy4()` which sets `ctx->mark` to `MARK_MAGIC_TO_PROXY`.
6. `netkit_run()` returns with `skb->mark` now set to `MARK_MAGIC_TO_PROXY`.
7. `netkit_xnet(skb)` immediately clears the mark before sending it on to the peer device and host namespace.

I think this then messes up routing to the proxy. Going from `NETKIT_SCRUB_BEFORE` -> `NETKIT_SCRUB_AFTER` basically swaps one problem for another. In the `veth` case `cil_from_container` runs on the ingress side after passing the packet to the primary device so any mark is preserved. I guess this is true for redirects as well. A program could mark a packet then call `bpf_redirect()`. If the redirect target expects the mark to be set, they'd be in for a surprise.

> NETKIT_SCRUB_NONE generally could be enough as an option anyway and if needed BPF prog could zero the data

Ack. `NETKIT_SCRUB_NONE` does seem to have the least surprising behavior when compared to `veth`. Hopefully there's no corner case where a container could set a mark to bypass ingress or egress policies.

**Comment by Anghille** [comment]
I am not sure if it is related; but l7 functionnality isnt working either. Using netkit (+BBR ba,dwith manager and native routing) and running a `cilium connectivity test` shows errors in all l7 related tests. 
The same config but switching to veth makes those same tests pass.

**Comment by borkmann** [comment]
@Anghille but this is with per-endpoint-routes config, right?

**Comment by Anghille** [comment]
It is with endpoint-routes disabled. Traffic flow by the cilium interface directly

**Comment by jrife** [comment]
@Anghille `cilium connectivity test` runs regularly as part of CI against netkit-enabled clusters. Can you share a bit of info about your configuration as well as the errors you encountered while running? What arguments did you use with `cilium connectivity test`? What does `cilium status` show?

**Comment by Anghille** [comment]
I can open a new ticket to avoid polluting this one with all informations necessary if you want :)

**Comment by jrife** [comment]
> I can open a new ticket to avoid polluting this one with all informations necessary if you want :)

Sure. Since you don't have endpoint routes enabled it sounds like this could be something different. If you have an easy way to reproduce your use case (preferably in a kind cluster) that would help as well.

**Comment by Anghille** [comment]
I'll do that right away !
